File: packages_mjs/app-yaml-config/src/validators.ts
/**
 * Validation schemas and custom errors.
 */

export class ValidationError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'ValidationError';
    }
}

export class ConfigNotInitializedError extends Error {
    constructor(message: string = "AppYamlConfig not initialized. Call initialize() first.") {
        super(message);
        this.name = 'ConfigNotInitializedError';
    }
}

export class ConfigAlreadyInitializedError extends Error {
    constructor(message: string = "AppYamlConfig already initialized.") {
        super(message);
        this.name = 'ConfigAlreadyInitializedError';
    }
}

export class ComputedKeyNotFoundError extends Error {
    constructor(key: string) {
        super(`Computed value '${key}' not defined`);
        this.name = 'ComputedKeyNotFoundError';
    }
}

export class CircularDependencyError extends Error {
    constructor(key: string, stack: string[]) {
        super(`Circular dependency detected for '${key}': ${stack.join(' -> ')}`);
        this.name = 'CircularDependencyError';
    }
}

File: packages_py/app_yaml_config/src/app_yaml_config/validators.py
"""Validation schemas and custom exceptions."""

class ValidationError(Exception):
    """Raised when validation fails (e.g., malformed YAML)."""
    pass

class ConfigNotInitializedError(Exception):
    """Raised when accessing config before initialize()."""
    pass

class ConfigAlreadyInitializedError(Exception):
    """Raised on second initialize() call."""
    pass

class ComputedKeyNotFoundError(Exception):
    """Raised when get_computed() key doesn't exist."""
    pass

class CircularDependencyError(Exception):
    """Raised when computed functions have circular dependencies."""
    pass
